go/apps/linux_host/linux_host.go

```diff
@@ -17,247 +17,534 @@ package main
-	hostSpec := flag.String("host_spec", "", "The spec to use for communicating with the parent (e.g., '/dev/tpm0')")
-	hostChannelType := flag.String("host_channel_type", "", "The type of the host channel (e.g., 'tpm', 'file', or 'unix')")
-	hostedProgramType := flag.String("hosted_program_type", "process", "The type of hosted program to create ('process', 'docker', or 'kvm_coreos')")
-	hostedProgramSocketPath := flag.String("hosted_program_socket_path", "linux_tao_host", "The directory in which to create unix sockets for hosted-program communication")
-
-	// QEMU/KVM CoreOS configuration with some reasonable defaults.
-	coreOSImage := flag.String("kvm_coreos_img", "coreos.img", "The path to a CoreOS image")
-	vmMemory := flag.Int("kvm_coreos_vm_memory", 1024, "The amount of RAM to give the VM")
-	sshFile := flag.String("kvm_coreos_ssh_auth_keys", "auth_ssh_coreos", "A path to the authorized keys file for SSH connections to the CoreOS guest")
-
-	// An action for the service to take.
-	action := flag.String("action", "start", "The action to take ('init', 'show', 'start', or 'stop')")
-	flag.Parse()
+var opts = []options.Option{
+	// Flags for all/most commands
+	{"tao_domain", "", "<dir>", "Tao domain configuration directory", "all"},
+	{"host", "", "<dir>", "Host configuration, relative to domain directory or absolute", "all"},
+	{"quiet", false, "", "Be more quiet", "all"},
+
+	// Flags for init (and start) command
+	{"root", false, "", "Create a root host, not backed by any parent Tao", "init,start"},
+	{"stacked", false, "", "Create a stacked host, backed by a parent Tao", "init,start"},
+	// TODO(kwalsh) hosted program type should be selectable at time of
+	// tao_launch. A single host should be able to host all types concurrently.
+	{"hosting", "", "<type>", "Hosted program type: process, docker, or kvm_coreos", "init"},
+	{"socket_dir", "", "<dir>", "Hosted program socket directory, relative to host directory or absolute", "init"},
+
+	// Flags for start command
+	{"foreground", false, "", "Run in the foreground", "start"},
+	// Using setsid (1) and shell redirection is an alternative -daemon:
+	//    sh$ setsid tao host start ... </dev/null >/dev/null 2>&1
+	//    sh$ setsid linux_host start ... </dev/null >/dev/null 2>&1
+	{"daemon", false, "", "Detach from tty, close stdio, and run as a daemon", "start"},
+
+	// Flags for root
+	{"pass", "", "<password>", "Host password for root hosts (for testing only!)", "root"},
+
+	// Flags for stacked
+	{"parent_type", "", "<type>", "Type of channel to parent Tao: TPM, pipe, file, or unix", "stacked"},
+	{"parent_spec", "", "<spec>", "Spec for channel to parent Tao", "stacked"},
+
+	// Flags for QEMU/KVM CoreOS init
+	{"kvm_coreos_img", "", "<path>", "Path to CoreOS.img file, relative to domain or absolute", "kvm"},
+	{"kvm_coreos_vm_memory", 0, "SIZE", "The amount of RAM (in KB) to give VM", "kvm"},
+	// TODO(kwalsh) shouldn't keys be generated randomly within the host?
```

Comment:
Yes. This should be marked as only for testing. Note that keys are generated randomly in the host to set up the state of the VM. These keys are currently used to deal with the fact that tao_launch doesn't have a way to request from the host that it start applications on KVM child linux_host programs. So, the scripts ssh into the child and call tao_launch there. This is bad.